dias/tracker.py

```diff
@@ -0,0 +1,400 @@
+
+        database = db
+
+        # file/analyzer pairs must be unique
+        indexes = ((("file_id", "analyzer_id"), True),)
+
+
+class Tracker:
+    """
+    Utility that keeps track of processed files for DIAS analyzers.
+
+    Interfaces with an SQLite database to assess which files have/haven't been already processed by various analyzers.
+    Creates the database, if it does not exist, and keeps it up-to-date with the current state of the staging directory.
+
+    Attributes
+    ----------
+    path : String
+        the location of the staging directory that the user is interested in
+
+    Methods
+    -------
+    __init__
+    create_tables
+    _get_filetypes
+    insert_files
+    add_analyzer_ine
+    remove_files
+    update_file_table
+    new_files
+    register_done
+    """
+
+    def __init__(
+        self,
+        path="/mnt/gong/staging",
+        file_tracking_db="/mnt/gong/staging/file_index.db",
+    ):
+        """Construct the Tracker client, and create SQLite tables, if they do not exist."""
+        self.base_path = path
+
+        self.logger = logging.getLogger(__name__)
+        self.logger.setLevel("INFO")
+
+        self.file_tracking_db = file_tracking_db
+        # create connection to sqlite database
+        db.init(os.path.expanduser(self.file_tracking_db))
+        try:
+            self.create_tables()
+        except OperationalError:
+            raise DiasUsageError(
+                "Unable to access file index db at location {0} for Tracker.".format(
+                    self.file_tracking_db
+                )
+            )
+
+    def create_tables(self):
+        """Create tables in database if they do not exist."""
+        file_types = {}
+
+        with db:
+
+            if not db.table_exists("file_type"):
+                self.logger.info("Creating file_type table")
+                db.create_tables([FileType])
+
+                for ft in FILETYPES:
+                    if not FileType.select().where(FileType.name == ft).exists():
+                        FileType.create(name=ft)
+
+            if not db.table_exists("analyzer"):
+                self.logger.info("Creating analyzer table")
+                db.create_tables([Analyzer])
+
+            if not db.table_exists("file"):
+                self.logger.info("Creating file table")
+                db.create_tables([File])
+
+                files = []
+                # grab a list of files on disk of FILETYPE
+                for ft in FILETYPES:
+                    files.extend(
+                        glob.glob(
+                            os.path.join(
+                                self.base_path, "*_{0}_corr".format(ft), "*.h5"
+                            )
+                        )
+                    )
+
+                self.insert_files(files)
+
+            if not db.table_exists("processed"):
+                self.logger.info("Creating processed table")
+                db.create_tables([Processed])
```

Comment:
Can't you wrap all of these into a single `db.create_tables([Processed, Analyzer...], safe=True)` call. I think `safe=True` is actually the default, but you would be able to skip all the `if not db.table_exists...` blocks as the db does it directly.

If you don't want to do that, I would suggest changing the calls from `db.create_tables([table])` to `db.create_table(table)`

Reply:
I don't want to do that at this moment!

The creation of some of them involve the actual population of their table, and for some of them, the already existence of a dependency table for that population. I only want to do the initial population once, if the table does not already exist. There are cleverer ways to do this, that would clean this up, and allow the batch table creation, but I don't think I have time to get that working. =( 

I will create an issue for that clean-up after deployment. I am going to change the calls for now.

Reply:
Nevermind, this was actually easier than I thought! Things have changed to make this easier.